### PR TITLE
feat(ci): introduce ESM and packages

### PR DESCRIPTION
**Pull request by heitorlessa**
*Issue #, if available:*

*Description of changes:*

Introduces ESM, NPM workspaces, and Biome as linter/formatter to allow for growth. Split up GitHub, markdown, and date code into separate workspace packages to ease testing later.

**Changes**

* [x] Move to `biome` as formatter/linter
* [x] Add linter to workspace
* [x] Update workflow to new path (use full path now for ESM)
* [x] Identify and fix bug of longest PRs not being available
* [x] Update existing issue if found over creating new ones
* [x] Split GitHub specific utilities into another package
* [x] Handle pagination since PRs and Issues are treated equal
* [x] Add which GH Action workflow triggered update if any
* [x] Create contract testing for GitHub utils
* [x] Add all properties GitHub accepts e.g., `assignees`
* [x] refactor: fail early and keep API as a single object
* [x] create Zod schema for IssueAsPR
* [x] Update search tests to remove org/repo that is unnecessary
* [x] Type all returns with new schemas and revisit docstring
* [x] Convert .js to .mjs
* [x] Update schema for Pull Request based on JSON Schema
* [x] Add docstring on purpose, notes, etc.
* [x] Experiment typing `core` with `actions/core` dev package
* [x] Expand `listPullRequests` parameters so we can fetch closed PRs too
* [x] Create an architecture to accommodate day-to-day automation and reporting
    * [x] Consolidate GitHub client logic into one package
    * [x] Create base issues and pull request models for extension
    * [x] Clean up imports by consolidating too fine-grained split
    * [x] [Optional] Sketch a formatter to format issues, pull requests, etc., output (`top feature requests`). No implementation need as we enumerate use cases and growth areas
    * [ ] [Optional] Create a GH Client interface
    * [x] Create base class for reports
    * [x] Create base class for templates
* [x] Create an actual builder to generate markdown table (and ease test assertion)
* [x] Move as models `TopFeatureRequests`, `TopMostCommented`
* [x] Clean up descriptions
* [x] Consolidate date_utils
* [x] Consolidate Markdown
* [x] Create a Markdown table builder to ease testing etc.
* [x] Create a GitHubActions
* [x] Recreate unit tests for list/create/etc.
* [x] Split reporting from report issue creation
* [x] Split out GitHub filters as conditions are growing
* [x] Support filtering on minimum creation date and max days without update (asked by @dreamorosi)

**Test**

* [x] Enforce fake timer in vitest for determinism
* [x] Create Zod Schema out of GitHub API to ease testing
* [x] Use `zod-mock` to create synthetically correct data for SDK fidelity
* [x] Override key mock data for accurate filtering _(e.g., labels)_
* [x] Create a builders
    * [x] Pull requests
    * [x] Issues
    * [x] Search results
* [x] Create network interceptors
    * [x] List Pull Requests
    * [x] Search Issues
    * [x] Create Issues
    * [x] Update Issues
    * [x] HTTP 500
* [x] Cleanup variable names to ensure testing can read as a story w/ GIVEN/WHEN/THEN
* [x] Unit test monthly roadmap
    * [x] Top Feature Requests (issues)
    * [x] Top Most Commented Issues
    * [x] Top Oldest Issues
    * [x] Longest Running PRs
    * [x] Create Monthly Roadmap Report
* [x] Clean up implementation by no longer depending on GH-Script as we've moved past at this stage
* [x] Replace org and repo with dummy values to avoid accidents in contract testing
* [x] Consolidate builders to reflect new `github` workspace

**Non-code related**

* [ ] Test from Python roadmap branch
* [x] Improve debugging with core group (test if concurrency doesn't affect nesting)
* [x] Create a new label `roadmap-report`
    * [x] Update code to check for existing issues with new label
    * [x] Update code to create new issue with this label
* [x] Fix any orphan npm workspace

## Questions to Andrea

* [x] **Design**: can we pass config objs to functions to avoid refactoring N places when new function properties are added? how to tackle it when we move to TS?
* [x] **Design**: fail early JS/TS design: do we fail early to avoid unnecessary I/O or propagate SDK errors for required arguments? e.g., updating issues require an issue number
* [x] **Tests**: how to structure testing (unit vs contract)?
* [x] **Tests**: should we define constants per `describe` or keep it global?
* [x] **Package**: need a deeper understanding on export implications

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by heitorlessa** [comment]
(no pressure on time) - Andrea.. when you have a moment, could you check the following I'm unsure before I work on making it more resilient?

1. Whether `biome` is setup correctly -- I noticed VSCode didn't fail when I tried to access a var not previously defined
2. Whether NPM workspaces is set correctly
3. Whether ESM for GH Actions Script is setup correctly (I'll test after merging)

File: package-lock.json
Comment (dreamorosi) [inline package-lock.json:1]: The lock file is missing some of the workspaces, this might be because they weren't initialized correctly (i.e. manually created the folder & added it to the `workspaces` section of the `package.json`).

When initializing a new workspace, do the following steps:
- create the folder (can be anywhere in the project - i.e. `packages/foo`)
- while in the project root, run `npm init -w packages/foo`

The last step will do three things:
- add `packages/foo` under `workspaces` in the `package.json`
- create a new `packages/foo/package.json` file
- add the workspace to the `package-lock.json` file - this includes a section that tells Node.js that this package has to be linked locally

--

To fix this delete the `package.json` and run `npm i` again, this will create a new lock file with the correct components.

Once that's done, you'll be able to run `npm ci` (`ci` stands fro clean install) which unlike `npm i` or `install` (alias) produces a deterministic install using the exact versions indicated in the lock file.

File: package.json [hunk 1]
```diff
@@ -1,9 +1,21 @@
...
   "devDependencies": {
+    "@biomejs/biome": "1.6.4",
+    "@octokit/rest": "^20.1.0",
+    "@types/github-script": "github:actions/github-script",
     "vitest": "^1.4.0"
   },
   "scripts": {
-    "test": "vitest"
+    "test": "vitest",
+    "lint": "biome lint packages --write",
```
Comment (dreamorosi) [inline package.json:16]: The `lint` command does not accept a `--write`, and in all cases I'd suggest to put the flag before the folders - although this is mainly to align with the docs and it'll work either way:

```json
"lint": "biome lint packages",
"format": "biome format --write packages",
```

After the change, running the `format` command causes the imports to be sorted so you might want to run and make another commit it before merging.

**Review by dreamorosi — COMMENTED** [review]
I have left a couple comments, one about the `biome` config and another about the npm workspaces, that are not configured correctly.

The GitHub script looks good (although I haven't run it myself).

About the linter not flagging the access of an `undefined` variable, I went to look at the rules and - perhaps understandably - there isn't any rule around this case specifically.

I said understandably because this seems like the type of check that only a proper language server can do. Essentially, even though you might have annotated said variable as potentially `undefined` with JSDocs comments, the linter is not able to understand the implications that this would have at runtime.

This type of check is the perfect example of what TypeScript would bring to the table, with the tradeoff of requiring a build step.

**Comment by heitorlessa** [comment]
Nice thanks a lot! 🫶. I'll experiment moving to TS, and if it's not too much a rewrite I'll push it 🍾

File: package-lock.json
Comment (heitorlessa) [reply package-lock.json:1]: aha TIL! I thought `npm init -w <workspace>` would initialize a new workspace, and not add OR update workspaces like scaffolding. 

Thank you, correcting that shortly.

File: package.json (hunk 1 above)
Comment (heitorlessa) [reply package.json]: fixed it :) tks!!

**Review by dreamorosi — COMMENTED** [review]
Check that the license field in the various `package.json` files is aligned with the one we set in the project (`MIT-0`).

**Comment by heitorlessa** [comment]
@dreamorosi I've added support for filtering issues and pull requests on creation date as well as minimum days without update.

I'm gonna merge this. I'll create separate issues to get your help on licensing, linting, and such.

Given the size of the PR, I'll create docs separately using `mkdocs`. The entire code has JSDoc along with `@example` tag for where it matters, and tests should be self-explanatory too.

I'll refrain from creating API docs as we will move to TS and I'm sure it'll be a different solution. I'll focus on user guide instead.

**THANK YOU** for your help at all times I was blocked :)
